2023/Q2/artifacts/PoS-quint/namada-redelegation.qnt

```diff
@@ -0,0 +1,735 @@
+                                                                                       validator.totalBonded.get(posState.epoch + PIPELINE_OFFSET) + amount))      
+        val updatedPos = posState.with("posAccount", posState.posAccount + amount)
+        {success: true, delegator: updatedDelegator, validator: updatedValidator, posState: updatedPos}
+      } else {
+        {success: false, delegator: delegator, validator: validator, posState: posState}
+      }
+    }
+
+    // The function unbond is called when a user wants to unbond tokens from a validator.
+    // 1. It first checks that the user has enough tokens bonded to the validator and that the validator is not frozen.
+    // 2. Then it uses iterateBondsUpToAmount to compute the set of bonds that must be unbonded.
+    // 3. It computes the set of unbond records to be added to the delegators unbonded variable.
+    // 4. It updates the delegator state by removing the unbonded bonds and adding the newly computed unbonds.
+    // 5. It applies any slashes to the unbonded bonds and stroes it in amountAfterSlashing.
+    // 6. It updates the validator's totalUnbonded.
+    // 7. It finally updates the validator's stale at the PIPELINE_OFFSET by subtracting amountAfterSlashing.
+
+    pure def unbond(delegator: DelegatorState, validatorAddress: Address, allValidators: Address -> ValidatorState, posState: PosState, amount: int, isRedelegation: bool): ResultUnbondTx = {
+      val validator = allValidators.get(validatorAddress)
+      val bonds = delegator.bonded.get(validator.address)
+      val redelegatedBonds = delegator.redelegatedBonded.get(validator.address)
+      val totalBonded = bonds.keys().fold(0, (sum, e) => sum + bonds.get(e))      
+      if (amount > 0 and amount <= totalBonded and validator.frozen < posState.epoch) {
+        // Compute the epoch at which the unbonded bonds will be withdrawable
+        val endEpoch = posState.epoch + PIPELINE_OFFSET + CUBIC_OFFSET + UNBONDING_OFFSET
+        // Compute the set of bonds to be unbonded
+        val resultUnbonding = iterateBondsUpToAmount(bonds, amount)
+        // Compute the updated delegators bonded variable by removing the unbonded bonds and adding a new bond in case one was partially unbonded                                                     
+        val updatedBonded = if (resultUnbonding.new.start == -1) bonds.mapRemoveSet(resultUnbonding.toRemove)
+                            else bonds.mapRemoveSet(resultUnbonding.toRemove).set(resultUnbonding.new.start, resultUnbonding.new.amount)
+        // Compute a modified redelegation bond in case one was partially unbonded
+        val modifiedRedelegation =  if (not(redelegatedBonds.has(resultUnbonding.new.start))) {epoch: -1, valsToRemove: Set(), valToModify: "", epochsToRemove: Set(), epochToModify: -1, newAmount: -1}
+                                    else computeModifiedRedelegation(redelegatedBonds, resultUnbonding.new.start, resultUnbonding.new.amount)
```

Comment:
should `redelegatedBonds` be `redelegatedBonds.get(resultUnbonding.new.start)`?

Reply:
Yes, thanks!